controllers/datafoundationclient.go

```diff
@@ -0,0 +1,226 @@
+	}
+	return r.reconcilePhases()
+}
+
+func (r *dataFoundationClientReconciler) initReconciler(reconciler *ManagedFusionOfferingReconciler, offering *v1alpha1.ManagedFusionOffering) {
+	r.ManagedFusionOfferingReconciler = reconciler
+	r.offering = offering
+
+	r.storageClient.Name = "ocs-storageclient"
+	r.storageClient.Namespace = offering.Namespace
+
+	r.defaultBlockStorageClassClaim.Name = "ocs-storagecluster-ceph-rbd"
+	r.defaultBlockStorageClassClaim.Namespace = offering.Namespace
+
+	r.defaultFSStorageClassClaim.Name = "ocs-storagecluster-cephfs"
+	r.defaultFSStorageClassClaim.Namespace = offering.Namespace
+}
+
+func (r *dataFoundationClientReconciler) parseSpec(offering *v1alpha1.ManagedFusionOffering) error {
+	r.Log.Info("Parsing ManagedFusionOffering Data Foundation Client spec")
+
+	isValid := true
+	onboardingTicket, found := offering.Spec.Config["onboardingTicket"]
+	if !found {
+		r.Log.Error(
+			fmt.Errorf("missing field: onboardingTicket"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+
+	providerEndpoint, found := offering.Spec.Config["dataFoundationProviderEndpoint"]
+	if !found {
+		r.Log.Error(
+			fmt.Errorf("missing field: dataFoundationProviderEndpoint"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+	if !isValid {
+		r.Log.Info("parsing ManagedFusionOffering Data Foundation Client spec failed")
+		return fmt.Errorf("invalid ManagedFusionOffering Data Foundation Client spec")
+	}
+	r.Log.Info("parsing ManagedFusionOffering Data Foundation Client spec completed successfuly")
+
+	r.dataFoundationClientSpec = dataFoundationClientSpec{
+		onboardingTicket: onboardingTicket,
+		providerEndpoint: providerEndpoint,
+	}
+	return nil
+}
+
+func (r *dataFoundationClientReconciler) reconcilePhases() (ctrl.Result, error) {
+	if !r.offering.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	} else {
+		if err := r.reconcileStorageClient(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileCSV(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileDefaultBlockStorageClassClaim(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileDefaultFSStorageClassClaim(); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+func (r *dataFoundationClientReconciler) reconcileStorageClient() error {
+	r.Log.Info("Reconciling StorageClient")
+
+	if r.dataFoundationClientSpec.providerEndpoint == "" {
+		return fmt.Errorf("dataFoundationProviderEndpoint is empty")
+	}
+
+	dfStorageProviderEndpointAsURL, err := url.ParseRequestURI(r.dataFoundationClientSpec.providerEndpoint)
+	if err != nil {
+		return err
+	}
+	if dfStorageProviderEndpointAsURL.Host == "" {
+		return fmt.Errorf("cannot retrieve host from %s, might be a malformed or partial URL", r.dataFoundationClientSpec.providerEndpoint)
```

Comment:
An error text should not start with a verb 
```suggestion
		return fmt.Errorf("Invalid provider endpoint %s, URL does not contain host", r.dataFoundationClientSpec.providerEndpoint)
```